### PR TITLE
fix issue of fp16 batch normalization

### PR DESCRIPTION
**Pull request by anaruse**
This PR is to allow you to use cuDNN for fp16 batch normalization.
Note, you should set **CHAINER_TYPE_CHECK=0**, when you test this branch.

### CODE REVIEW

**Comment by unnonouno** [comment]
Why not determine dtype based on input variable?

**Comment by anaruse** [comment]
Thank you for your question, but sorry I don't understand exactly what you are asking. Is it a question to spec. of cuDNN (more specifically, cudnnBatchNormalizationForwardTrainig) ?

**Comment by unnonouno** [comment]
Sorry for late reply. I feel you can select `dtype_param` based on `dtype_bn`. Why do you need to add an argument `dtype_param` in `__init__`?

**Comment by anaruse** [comment]
I suppose the `__init__()` that you mentioned is an `__init__()` of `class BatchNormalizationFunction(function.Function)`, correct?
Then, looks you're right. I thought the dtype_param must be passed to the `__init__()` as an argument, but the dtype can be obtained from variables like gamma and beta in a `forward()`. I will fix the code in that way.

**Comment by unnonouno** [comment]
Yes > `BatchNormalizationFunction`
Thank you!

**Comment by anaruse** [comment]
The argument dtype_param has been removed from the `__init__()` of `BatchNormalizationFunction`. Please check.

File: chainer/links/normalization/batch_normalization.py [hunk 1]
```diff
@@ -66,22 +66,28 @@ def __init__(self, size, decay=0.9, eps=2e-5, dtype=numpy.float32,
                  use_gamma=True, use_beta=True,
                  initial_gamma=None, initial_beta=None, use_cudnn=True):
         super(BatchNormalization, self).__init__()
+        # [cuDNN] dtype of prameters like gamma, beta, etc. must be float32,
+        # even when dtype of input/output tensors are float16.
+        dtype_param = dtype
+        if use_cudnn and dtype_param == numpy.float16:
+            dtype_param = numpy.float32
```
Comment (unnonouno) [inline chainer/links/normalization/batch_normalization.py]: Oh, i think the current type-check code of batch-normalization function forbids this type of argument. See https://github.com/pfnet/chainer/pull/2299/files#diff-148edea535bb806260c487ca31869a91R64

File: chainer/functions/normalization/batch_normalization.py [hunk 2]
```diff
@@ -77,6 +77,10 @@ def check_type_forward(self, in_types):
     def forward(self, inputs):
         xp = cuda.get_array_module(*inputs)
         x, gamma, beta = inputs[:3]
+        if gamma.dtype != beta.dtype:
```
Comment (anaruse) [inline chainer/functions/normalization/batch_normalization.py]: Agree. The dtype checking is not necessary here. Will remove.

File: chainer/links/normalization/batch_normalization.py (hunk 1 above)
Comment (anaruse) [inline chainer/links/normalization/batch_normalization.py]: I'm wondering how to change testing codes...

**Comment by anaruse** [comment]
I've just rebased the branch to the latest master, and all failures of nosetests on batch normalization has disappeared. This is weird, because there are data type difference in input arguments of method `forward()` in `BatchNormalizationFunction` when fp16 is used, and the type checking method `check_type_forward()` does not allow such difference.  Seems some fp16 use cases are not covered by [a test code for batch normalization](https://github.com/anaruse/chainer/blob/fp16_BN/tests/chainer_tests/functions_tests/normalization_tests/test_batch_normalization.py)

**Comment by anaruse** [comment]
My previous comment was wrong.

I did run only `chainer_tests/functions_tests/normalization_tests` but not `chainer_tests/links_tests/normalization_tests`. Actually there were a lot of failures with `chainer_tests/links_tests/normalization_tests` due to fp16...

To appropriately support fp16 with cuDNN, I've added various modifications to the link, function and testing codes of batch normalization. The most important fix is to extract a processing of parameters' initialization from a mathod `__init__()` of class BatchNormalization and make a new method `init_params()`. For more details, refer to a comment in the method `init_params()`.

**Comment by unnonouno** [comment]
I think when `__init__` is called with `dtype=numpy.float16`, we don't need to support other types of inputs in `__call__`. In the original implementation, different types of inputs are not permitted.

File: chainer/functions/normalization/batch_normalization.py [hunk 3]
```diff
@@ -56,27 +56,38 @@ def check_type_forward(self, in_types):
...
-        type_check.expect(
-            x_type.dtype.kind == 'f',
-            x_type.ndim >= gamma_type.ndim + 1,
-            x_type.shape[1:1 + M] == gamma_type.shape,
-            # TODO(beam2d): Check shape
-            gamma_type.dtype == x_type.dtype,
-            beta_type.dtype == x_type.dtype,
-            gamma_type.shape == beta_type.shape,
-        )
+        if x_type.dtype.eval() == numpy.float16:
```
Comment (unnonouno) [inline chainer/functions/normalization/batch_normalization.py]: Arguments of `expect` represents AND. So, you can split them.
```
if xp is not numpy and cuda.cudnn_enabled and self.use_cudnn and \
     self.cudnn_dim_ok and _cudnn_version >= 5000 and \
     x_type.dtype.eval() == numpy.float16:
  type_check.expect(
    gamma_type.dtype == numpy.float32,
    beta_type.dtype == numpy.float32,
   )
else:
  type_check.expect(
    gamma_type.dtype == x_type.dtype,
    beta_type.dtype == x_type.dtype,
   )
```

Comment (anaruse) [inline chainer/functions/normalization/batch_normalization.py]: I will fix the code based on your suggestion.
BTW, I think that whether x is `numpy` or `cupy` is not easily determined in the method `check_type_forward`, because the argument is `in_types` and not `inputs`...  Any idea?

**Comment by anaruse** [comment]
The question is data type of parameters like gamma and beta.

When an `__init__()` of a class `BatchNormalization` is called with `dtype=numpy.float16` and `use_cudnn=True`, date type of parameters can be either `float32` or `float16` depending on a shape (number of dimensions) of x.  For example, when number of dimensions of x is two or four that is supported by cuDNN, data type of parameters must be `float32` (in order to use cuDNN). But the data type must be `float16` in other cases. That is why I split the original `__init__()` into a new `__init__()` and `init_params()`.

Honestly speaking, I don't think this is good and readable approach...  Any idea?

File: chainer/functions/normalization/batch_normalization.py (hunk 3 above)
Comment (unnonouno) [inline chainer/functions/normalization/batch_normalization.py]: Oh, that's right...
